### PR TITLE
feat(upload folder): init web and node upload folder with manifest implementations PE-4643

### PR DESCRIPTION
**Pull request by fedellen**
This PR adds a basic `uploadFolder` method with both Node and Web implementations. For now it is just a wrapper around the `uploadFile` in a sense. On nodeJS it will send each file found within the folder path provided. On Web we will accept an array of `File` type provided by the browser. After uploading we optionally create a manifest of the upload. 

There are more features we could build onto here in the future including BDI support, multipart-uploads, and ephemeral signing on web to prevent signature authorization of each item.

### CODE REVIEW

File: src/node/upload.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        fileStreamFactory: () => createReadStream(absoluteFilePath),
+        fileSizeFactory: () => statSync(absoluteFilePath).size,
+        signal,
+        dataItemOpts,
+      });
+
+      fileResponses.push(result);
+      const relativePath = absoluteFilePath.replace(folderPath, '');
+      paths[relativePath] = { id: result.id };
+    }
```
Comment (dtfiedler) [inline src/node/upload.ts:89]: a few comments:
- consider doing this concurrently and in a p-limit
- consider how we want to handle one file falling after x number of retries. we may want to provide the option to abort on failure, e.g. `abortOnFailure` or `throwOnFailure` that throws the error of the last upload. alternatively, if `throwOnFailure` is `false` we catch any errors and return them with the result here. I'm guessing we do something similar in ardrive-js?

File: src/types.ts [hunk 2]
```diff
@@ -126,6 +126,43 @@ export type TurboUploadDataItemResponse = {
...
+  manifest: ArweaveManifest;
+};
+
+export type ArweaveManifest = {
+  manifest: 'arweave/paths';
+  version: '0.2.0';
+  index: { path: string };
+  paths: Record<string, { id: string }>;
+  fallback?: { id: string };
+};
```
Comment (dtfiedler) [inline src/types.ts:164]: is this available anywhere else?

Comment (fedellen) [reply src/types.ts:164]: not sure, we have one in ardrive-core-js for sure.

File: src/node/upload.ts (hunk 1 above)
Comment (fedellen) [reply src/node/upload.ts]: 110a424 got the pLimit part worked out here

File: src/common/factory.ts [hunk 3]
```diff
@@ -55,6 +54,10 @@ export class TurboBaseFactory {
     this.logger.setLogFormat(format);
   }
 
+  private static isWeb() {
+    return typeof window !== 'undefined';
+  }
```
[the remark this answers is not in the document]
Comment (fedellen) [reply src/common/factory.ts]: 038c098

File: src/node/upload.ts (hunk 1 above)
Comment (fedellen) [reply src/node/upload.ts]: a258aa6 heres is a throw on failure commit, with errors being optionally returned on false

File: src/node/upload.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        const relativePath = absoluteFilePath.replace(folderPath + '/', '');
+        paths[relativePath] = { id: result.id };
+      } catch (error) {
+        if (throwOnFailure) {
+          throw error;
+        }
+        this.logger.error(`Error uploading file: ${absoluteFilePath}`);
+        this.logger.error(error);
+        errors.push(error);
+      }
```
Comment (dtfiedler) [inline src/node/upload.ts:119]: i've started to become a fan of the await/catch pattern - particularly in these nested try blocks. what do you think?

```suggestion
        const result = await this.uploadFile({
          fileStreamFactory: () => createReadStream(absoluteFilePath),
          fileSizeFactory: () => statSync(absoluteFilePath).size,
          signal,
          dataItemOpts: {
            ...dataItemOpts,
            tags: [
              ...(dataItemOpts?.tags ?? []),
              { name: 'Content-Type', value: contentType },
            ],
          },
        }).catch((error: any) => {
             this.logger.error(`Error uploading file: ${absoluteFilePath}`, {
                  message: error?.message,
                  stack: error?.stack
             });
             if (throwOnFailure) {
                throw error;
             }
             errors.push(error);
        });
        fileResponses.push(result);
        const relativePath = absoluteFilePath.replace(folderPath + '/', '');
        paths[relativePath] = { id: result.id };
```

Comment (fedellen) [reply src/node/upload.ts]: that gives some issues on types 
image

we need it to be a then chain on the other parts. and then we're mixing await and then a bunch, might as well make this a then.catch function

Comment (fedellen) [reply src/node/upload.ts:119]: I converted to then/catch, but the linter actually wanted me to convert it back to async function with try/catch. just reverted this for now. overall I think mixing `await` with `.catch()` makes flow control a bit tougher to reason about

File: src/web/upload.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+        const relativePath = file.name ?? file.webkitRelativePath;
+        paths[relativePath] = { id: result.id };
+        fileResponses.push(result);
+      } catch (error) {
+        if (throwOnFailure) {
+          throw error;
+        }
+        this.logger.error(`Error uploading file: ${file.name}`);
+        this.logger.error(error);
```
Comment (dtfiedler) [inline src/web/upload.ts:89]:
```suggestion
        this.logger.error(`Error uploading file: ${file.name}`, error);
```

File: src/common/factory.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (fedellen) [reply src/common/factory.ts]: the signer change could probably be a separate branch/ticket. we could maintain backwards comp by being mindful of our exports. but that signer out of the box doesn't provide signing of arbitrary transactions or public key access that we require in some flows here

**Review by dtfiedler — APPROVED** [review]
looks good! :shipit:

**Pull request merged**